File: browser/src/functions/elements.rs
//! `browser::elements` — the page's visible, enabled controls as an indexed
//! table with `n` refs, plus the visible text, from one in-page evaluation
//! (`elements.js`). Also the in-page input guard every ref-addressed
//! `browser::act` runs before dispatching input, and the short post-input
//! settle wait. Ported from browser-use/jev-ultrafast.

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// Reads the table; evaluates to `ElementsOutput` JSON (without
/// `generation`), or null while the document has no body.
pub const OBSERVE_JS: &str = include_str!("elements.js");

/// Most elements one table lists (`elements.js` enforces it; the rest are
/// counted in `omitted`). Also keeps every judge `choice` under its 255 cap.
pub const MAX_ELEMENTS: usize = 250;

// ponytail: the registry lives on `window`, so the page can see or tamper
// with it; move observation into an isolated world (Page.createIsolatedWorld)
// if a hostile page ever matters.

/// Runs with `this` = the target element (`Runtime.callFunctionOn`). Scrolls
/// it into view when it is outside the viewport. For `click`, `type` and
/// `select` it then refuses a detached, disabled, hidden, read-only (type) or
/// covered element, so input never lands on whatever sits on top of it.
/// `type` focuses and, for `<input>`/`<textarea>`, selects the current value
/// so the inserted text replaces it. `select` picks the enabled option whose
/// value or label matches and fires `input`/`change`. Returns `{error}` on
/// refusal.
pub const GUARD_FN: &str = r#"function (kind, option) {
  const e = this;
  if (!e.isConnected) return { error: 'the element is no longer in the page' };
  const doc = e.ownerDocument, win = doc.defaultView;
  let r = e.getBoundingClientRect();
  if (r.bottom <= 0 || r.right <= 0 || r.top >= win.innerHeight || r.left >= win.innerWidth) {
    e.scrollIntoView({ block: 'center', inline: 'nearest' });
    r = e.getBoundingClientRect();
  }
  if (kind !== 'click' && kind !== 'type' && kind !== 'select') return {};
  if (e.matches(':disabled') || e.closest('[aria-disabled="true"],[inert]')) return { error: 'the element is disabled' };
  if (typeof e.checkVisibility === 'function' && !e.checkVisibility({ checkVisibilityCSS: true })) return { error: 'the element is hidden' };
  if (kind === 'type' && (e.readOnly || e.getAttribute('aria-readonly') === 'true')) return { error: 'the element is read-only' };
  if (r.width > 0 && r.height > 0) {
    const hit = doc.elementFromPoint(r.x + r.width / 2, r.y + r.height / 2);
    const own = hit && (e.contains(hit) || hit.contains(e) || [...(e.labels || [])].some((l) => l.contains(hit)));
    if (!own) {
      const what = !hit ? 'nothing (outside the viewport)' : hit.tagName.toLowerCase() + (hit.id ? '#' + hit.id : '') +
        (typeof hit.className === 'string' && hit.className.trim() ? '.' + hit.className.trim().split(/\s+/)[0] : '');
      return { error: 'the element is covered by ' + what };
    }
  }
  if (kind === 'type') {
    e.focus();
    if ((e.tagName === 'INPUT' || e.tagName === 'TEXTAREA') && typeof e.select === 'function') {
      e.select();
      return { replaced: true };
    }
    return {};
  }
  if (kind === 'select') {
    if (e.tagName !== 'SELECT') return { error: 'select needs a <select> element; for a custom dropdown, click it and then click the option' };
    const want = String(option ?? '').trim();
    const enabled = [...e.options].filter((o) => !o.disabled && !o.closest('optgroup[disabled]'));
    const o = enabled.find((o) => o.value === want || o.label.trim() === want) ||
      enabled.find((o) => o.label.trim().toLowerCase() === want.toLowerCase());
    if (!o) return { error: 'no enabled option matches ' + JSON.stringify(want) + '; options: ' + enabled.slice(0, 20).map((o) => o.label.trim()).join(', ') };
    if (e.multiple) o.selected = true; else e.value = o.value;
    e.dispatchEvent(new Event('input', { bubbles: true }));
    e.dispatchEvent(new Event('change', { bubbles: true }));
    return { selected: o.label.trim() };
  }
  return {};
}"#;

/// Promise that resolves once the page had a chance to react to input: two
/// animation frames, capped at 50 ms. After typing into an ARIA combobox it
/// instead waits for a visible `[role=option]`, capped at 200 ms, so the
/// next read sees the suggestions.
pub fn settle_script(typed: bool) -> String {
    format!(
        r#"new Promise((resolve) => {{
  const a = document.activeElement;
  const combo = {typed} && !!a && a.getAttribute('role') === 'combobox';
  let frames = 0, done = false;
  const finish = () => {{ if (!done) {{ done = true; resolve(true); }} }};
  setTimeout(finish, combo ? 200 : 50);
  const options = () => {{
    const ids = (a.getAttribute('aria-controls') || a.getAttribute('aria-owns') || '').split(/\s+/).filter(Boolean);
    const roots = ids.length ? ids.map((id) => document.getElementById(id)).filter(Boolean) : [document];
    return roots.flatMap((root) => [...root.querySelectorAll('[role="option"]')]).some((o) => {{
      const r = o.getBoundingClientRect();
      return r.width && r.height && r.bottom > 0 && r.top < innerHeight && o.checkVisibility({{ checkVisibilityCSS: true }});
    }});
  }};
  const tick = () => {{
    if (done) return;
    if (++frames >= 2 && (!combo || options())) finish(); else requestAnimationFrame(tick);
  }};
  requestAnimationFrame(tick);
}})"#
    )
}

/// The registry id behind an `n` ref (`n12` → 12), or None for other refs.
pub fn registry_id(r: &str) -> Option<u64> {
    r.strip_prefix('n')?.parse().ok()
}

#[derive(Debug, Deserialize, JsonSchema)]
pub struct ElementsInput {
    pub session_id: String,
}

/// One control. `operations` says what `browser::act` can do with it.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct Element {
    /// `n` ref `browser::act` accepts; valid while the element stays in this
    /// document.
    #[serde(rename = "ref")]
    pub r#ref: String,
    pub role: String,
    /// Accessible name (aria-labelledby, aria-label, label, text, title, or
    /// placeholder).
    pub label: String,
    /// Current value of a field or dropdown; a set password reads `(set)`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub value: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub checked: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub selected: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expanded: Option<String>,
    /// `click`, `type` and/or `select`.
    pub operations: Vec<String>,
    /// Enabled `<select>` option labels (first 50), for `act` `select`.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub options: Vec<String>,
    /// Options beyond the listed 50; `select` still accepts them by label.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub more_options: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema)]
pub struct ElementsOutput {
    pub url: String,
    pub title: String,
    /// Visible text in the viewport, up to 6000 characters.
    pub text: String,
    pub can_scroll_up: bool,
    pub can_scroll_down: bool,
    /// Visible, enabled controls inside the viewport, in document order.
    pub elements: Vec<Element>,
    /// Controls past the 250-element cap, not listed.
    pub omitted: u32,
    /// Document generation; navigation advances it and every `n` ref dies.
    #[serde(default)]
    pub generation: u64,
}

impl ElementsOutput {
    /// The same page for progress purposes: document, text and controls.
    pub fn same_page(&self, other: &Self) -> bool {
        self.url == other.url
            && self.text == other.text
            && self.elements == other.elements
            && self.can_scroll_up == other.can_scroll_up
            && self.can_scroll_down == other.can_scroll_down
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registry_ids_come_only_from_n_refs() {
        assert_eq!(registry_id("n12"), Some(12));
        assert_eq!(registry_id("e12"), None);
        assert_eq!(registry_id("n"), None);
        assert_eq!(registry_id("nx"), None);
    }

    #[test]
    fn page_json_deserializes_without_generation() {
        let out: ElementsOutput = serde_json::from_value(serde_json::json!({
            "url": "https://x.test/", "title": "t", "text": "hello",
            "can_scroll_up": false, "can_scroll_down": true, "omitted": 0,
            "elements": [{"ref": "n1", "role": "button", "label": "Save", "operations": ["click"]}]
        }))
        .unwrap();
        assert_eq!(out.elements[0].r#ref, "n1");
        assert_eq!(out.generation, 0);
        assert!(out.same_page(&out.clone()));
    }
}

File: browser/src/functions/elements.js
// browser::elements — the page's visible, enabled controls in the viewport as
// an indexed table, plus the visible text, read in one evaluation. Ported from
// browser-use/jev-ultrafast's snapshot.js. Each element gets a code-owned `n`
// ref: a WeakMap keeps one id per real DOM node for the document's lifetime and
// a Map keeps the live node the ref resolves to, so a ref never names a
// different element. A new document's registry numbers from `first`, above
// every id the session handed out before, so a ref read on an earlier page
// never resolves on a later one. Worker overlays (`iii-*` ids) are never
// listed. Called with `first`; returns `next` for the session to remember.
((first) => {
  if (!document.body) return null;
  const R = (window.__iiiElements ||= { ids: new WeakMap(), nodes: new Map(), next: first });
  const identity = (e) => {
    if (!R.ids.has(e)) R.ids.set(e, R.next++);
    const id = R.ids.get(e);
    R.nodes.set(id, e);
    return id;
  };
  for (const [id, e] of R.nodes) if (!e.isConnected) R.nodes.delete(id);
  const cap = (s, n) => {
    s = String(s ?? '').replace(/\s+/g, ' ').trim();
    return s.length > n ? s.slice(0, n - 1) + '…' : s;
  };
  const hidden = (e) =>
    !!e.closest('[aria-hidden="true"],[inert],[id^="iii-"]') ||
    !e.checkVisibility({ checkVisibilityCSS: true });
  const name = (e, seen = new Set()) => {
    if (!e || seen.has(e)) return '';
    seen.add(e);
    const referenced = (e.getAttribute('aria-labelledby') || '')
      .split(/\s+/)
      .map((id) => name(document.getElementById(id), seen))
      .filter(Boolean)
      .join(' ');
    return (
      referenced ||
      e.getAttribute('aria-label') ||
      [...(e.labels || [])].map((l) => name(l, seen)).filter(Boolean).join(' ') ||
      (['button', 'submit', 'reset'].includes(e.type) ? e.value : '') ||
      e.getAttribute('alt') ||
      (e.tagName === 'INPUT' || e.tagName === 'SELECT' || e.tagName === 'TEXTAREA'
        ? ''
        : [...e.childNodes]
            .map((n) =>
              n.nodeType === 3
                ? n.textContent
                : n.nodeType === 1 && n.getAttribute('aria-hidden') !== 'true'
                  ? name(n, seen)
                  : '',
            )
            .join(' ')
            .trim()) ||
      e.getAttribute('title') ||
      e.getAttribute('placeholder') ||
      ''
    );
  };
  const roles = ['button', 'link', 'checkbox', 'radio', 'switch', 'tab', 'menuitem',
    'menuitemcheckbox', 'menuitemradio', 'option', 'gridcell', 'combobox', 'textbox',
    'searchbox', 'spinbutton', 'slider'];
  const selector =
    'a[href],button,input,textarea,select,summary,[contenteditable]:not([contenteditable="false"]),' +
    roles.map((role) => '[role="' + role + '"]').join(',');
  const role = (e) => {
    const explicit = e.getAttribute('role');
    if (roles.includes(explicit)) return explicit;
    if (e.tagName === 'BUTTON' || e.tagName === 'SUMMARY') return 'button';
    if (e.tagName === 'A') return 'link';
    if (e.tagName === 'SELECT') return 'combobox';
    if (e.tagName === 'TEXTAREA' || e.isContentEditable) return 'textbox';
    if (e.tagName === 'INPUT') {
      if (['checkbox', 'radio'].includes(e.type)) return e.type;
      if (['button', 'submit', 'reset', 'image'].includes(e.type)) return 'button';
      if (e.type === 'search') return 'searchbox';
      if (e.type === 'number') return 'spinbutton';
      if (['text', 'email', 'url', 'tel', 'password'].includes(e.type)) return 'textbox';
    }
    return null;
  };
  const elements = [];
  let omitted = 0;
  for (const e of document.querySelectorAll(selector)) {
    if (e.tagName === 'INPUT' && ['file', 'hidden'].includes(e.type)) continue;
    if (e.matches(':disabled') || e.closest('[aria-disabled="true"]') || hidden(e)) continue;
    const rname = role(e);
    const r = e.getBoundingClientRect();
    const x = r.x + r.width / 2;
    const y = r.y + r.height / 2;
    if (!rname || r.width <= 0 || r.height <= 0 || x < 0 || y < 0 || x >= innerWidth || y >= innerHeight) continue;
    if (rname === 'gridcell' && e.querySelector('button,[role="button"]')) continue;
    if (elements.length >= 250) {
      omitted++;
      continue;
    }
    const el = { ref: 'n' + identity(e), role: rname, label: cap(name(e) || rname, 120), operations: [] };
    for (const key of ['checked', 'selected', 'expanded']) {
      const value = e.getAttribute('aria-' + key);
      if (value !== null) el[key] = value;
    }
    if (['checkbox', 'radio'].includes(e.type)) el.checked = String(e.checked);
    // What the field is, beyond its label: lets a caller's `inputs` key such
    // as `password` or `email` find the field whatever language labels it.
    if (e.tagName === 'INPUT' || e.tagName === 'TEXTAREA') {
      if (e.tagName === 'INPUT' && e.type && e.type !== 'text') el.input_type = e.type;
      const fieldName = e.getAttribute('name') || e.id;
      if (fieldName) el.name = cap(fieldName, 60);
      const auto = e.getAttribute('autocomplete');
      if (auto && auto !== 'on' && auto !== 'off') el.autocomplete = cap(auto, 60);
    }
    if (e.tagName === 'SELECT') {
      el.operations.push('select');
      el.value = cap([...e.selectedOptions].map((o) => o.label).join(', '), 120);
      const options = [...e.options].filter((o) => !o.disabled && !o.closest('optgroup[disabled]'));
      el.options = options.slice(0, 50).map((o) => cap(o.label, 80));
      if (options.length > 50) el.more_options = options.length - 50;
    } else {
      const editable =
        !e.readOnly &&
        e.getAttribute('aria-readonly') !== 'true' &&
        (['textbox', 'searchbox', 'spinbutton'].includes(rname) ||
          (rname === 'combobox' && ['INPUT', 'TEXTAREA'].includes(e.tagName)));
      if (editable) el.operations.push('type');
      el.operations.push('click');
      if (e.type === 'password') el.value = e.value ? '(set)' : '';
      else if (editable || rname === 'combobox')
        el.value = cap('value' in e ? e.value : e.innerText, 120);
      if (el.value === '') delete el.value;
    }
    elements.push(el);
  }
  const words = [];
  const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_TEXT);
  const range = document.createRange();
  let node;
  let length = 0;
  while ((node = walker.nextNode()) && length < 6000) {
    const value = node.textContent.replace(/\s+/g, ' ').trim();
    const parent = node.parentElement;
    if (!value || !parent || parent.closest('script,style,noscript,template') || hidden(parent)) continue;
    range.selectNodeContents(node);
    const r = range.getBoundingClientRect();
    if (r.width > 0 && r.height > 0 && r.bottom > 0 && r.top < innerHeight && r.right > 0 && r.left < innerWidth) {
      words.push(value);
      length += value.length + 1;
    }
  }
  const height = document.documentElement.scrollHeight;
  return {
    url: location.href,
    title: document.title,
    // Still loading: the document itself, or a region the page marks busy.
    busy: document.readyState !== 'complete' || !!document.querySelector('[aria-busy="true"]'),
    text: words.join('\n').slice(0, 6000),
    can_scroll_up: scrollY > 0,
    can_scroll_down: scrollY + innerHeight < height - 2,
    elements,
    omitted,
    next: R.next,
  };
})
